### DISCUSSION TITLE
Console for credential management

### DISCUSSION THREAD

**Discussion by sukhpreetssekhon** [Sunbird-RC/community#332] [Ideas & Enhancements]
A state education department plans to leverage RC to issue educational credentials (proof of enrolment, proof of scholarship / aid, proof of learning achievement). The state has existing transaction systems which will be leveraged for generating credentials. Use cases for a credential management console include:

1. Defining credential schema and templates
2. Issuance of credentials from existing state transaction systems
3. Update of credentials owing to some update in the linked transaction system
4. Revocation of credentials due to deletion / update of record in transaction system
5. Triggering notifications (via an SMS gateway) to learners on issuance / update / revocation of a credential. 

Are there any implementations available in RC or in the larger open source ecosystem that can be contributed back to RC? 

@parthlawate @coolbung @girirajdaga - Based on our past discussions will be good if you can add perspective on what is available in RC or as open source assets that can be contributed back to RC in the next few weeks.

cc - @suresh12 @chakshugautam @surendrasinghs

**Comment by parthlawate** [comment]
@sukhpreetsamagra Super exciting usecase ! Sharing possible mappings I see below against each of your use cases 

1. Defining credential schema and templates

> In Progress project of Sunbird RC Admin portal allows for creating new schemas & templates (for any kind of registries)
> https://github.com/Sunbird-RC/Admin-Portal
> 

2. Issuance of credentials from existing state transaction systems

> Issuance portal . ready reference portal available.  
> https://docs.sunbirdrc.dev/sample-use-cases/certificate-issuance 
> https://github.com/Sunbird-RC/vc-issuance 
> 
> We can either use the portal /expose it to them with the UI & SSO with state systems OR just make the Service for Issuance available to the states
> 

3. Update of credentials owing to some update in the linked transaction system
> From an RC perspective, this would mean revocation of the credential & Issuance of a new credential.  

4, Revocation of credentials due to deletion / update of record in transaction system
> same as above

5. Triggering notifications (via an SMS gateway) to learners on issuance / update / revocation of a credential.
>Right now this isnt there in core.. but we can integrate an opensource service for this? We've used this in some projects https://github.com/notifme/notifme-sdk 

We also have a pretty mature PHP based implementation.. but that wont work in this case i think. But check it out here for ideas https://github.com/techjoomla/TJ-Notifications

.. Could UCI be integrated ?

**Reply by suresh12** [reply]
> @sukhpreetsamagra Super exciting usecase ! Sharing possible mappings I see below against each of your use cases
> > @parthlawate comments below. 
> 
> 1. Defining credential schema and templates
> 
> > In Progress project of Sunbird RC Admin portal allows for creating new schemas & templates (for any kind of registries)
> > https://github.com/Sunbird-RC/Admin-Portal
> > > @parthlawate We might have to extend the Admin Portal to include credential schema and template management described as part of the discussion https://github.com/Sunbird-RC/community/discussions/319 
> 
> 2. Issuance of credentials from existing state transaction systems
> 
> > Issuance portal . ready reference portal available.
> > https://docs.sunbirdrc.dev/sample-use-cases/certificate-issuance
> > https://github.com/Sunbird-RC/vc-issuance
> > We can either use the portal /expose it to them with the UI & SSO with state systems OR just make the Service for Issuance available to the states
> > > @parthlawate We should be able to extend Issuance Portal with SSO and other necessary changes to support Credential Management use cases.
> 
> 3. Update of credentials owing to some update in the linked transaction system
> 
> > From an RC perspective, this would mean revocation of the credential & Issuance of a new credential.
> > > This should be supported using the Issuance Portal.
> 
> 4, Revocation of credentials due to deletion / update of record in transaction system
> 
> > same as above
> > > This should be supported using the Issuance Portal.
> 
> 5. Triggering notifications (via an SMS gateway) to learners on issuance / update / revocation of a credential.
> 
> > Right now this isnt there in core.. but we can integrate an opensource service for this? We've used this in some projects https://github.com/notifme/notifme-sdk
> 
> We also have a pretty mature PHP based implementation.. but that wont work in this case i think. But check it out here for ideas https://github.com/techjoomla/TJ-Notifications
> 
> .. Could UCI be integrated ?
> > @parthlawate We can use UCI for notifications.

**Comment by pramodkvarma** [comment]
Let us make sure all external integrations from RC are loosely coupled and well interfaced out (cQube, UCI, etc.).

**Comment by girirajdaga** [comment]
> A state education department plans to leverage RC to issue educational credentials (proof of enrolment, proof of scholarship / aid, proof of learning achievement). The state has existing transaction systems which will be leveraged for generating credentials. Use cases for a credential management console include:
> 
> 1. Defining credential schema and templates
> 2. Issuance of credentials from existing state transaction systems
> 3. Update of credentials owing to some update in the linked transaction system
> 4. Revocation of credentials due to deletion / update of record in transaction system
> 5. Triggering notifications (via an SMS gateway) to learners on issuance / update / revocation of a credential.
> 
> Are there any implementations available in RC or in the larger open source ecosystem that can be contributed back to RC?
> 
> @parthlawate @coolbung @girirajdaga - Based on our past discussions will be good if you can add perspective on what is available in RC or as open source assets that can be contributed back to RC in the next few weeks.
> 
> cc - @suresh12 @ChakshuGautam @surendrasinghs

Affinidi Developer tools can be used as building blocks and integrated with application logic for transaction management. They are available [here](https://github.com/affinidi/affinidi-core-sdk) and there is a sample reference app also available that illustrates how to use it https://github.com/affinidi/elements-reference-app-frontend. 1-4 can be easily built using these building blocks. There is a managed solution with Affinidi for #5. 

cc: @sukhpreetsamagra @suresh12 @ChakshuGautam @surendrasinghs

**Comment by amarts** [comment]
> Are there any implementations available in RC or in the larger open source ecosystem that can be contributed back to RC?

@dhiway provides the solution through 2 forms. 

### 1. Opensource
A developer can use combination of below repos (for asks 1-4)
 * https://github.com/dhiway/cord (public ledger infra), 
 * https://github.com/dhiway/cord.js (Typescript based SDK to interact with Ledger, and also provide VC / Schema management methods), 
 * https://github.com/dhiway/cord-agent-app-v1 (example typescript based application which implements REST APIs. for 1-4 in the ask above).

Happy to address the gaps in documentation, if someone fails to understand something in above projects.
  
### 2. Commercial
  
Using the same project as the underlaying ledger, (ie, cord / cord.js) but with few more APIs and auth layer, and few more options to create templates for given schema etc, Dhiway provides https://studio.dhiway.com as a SaaS product, which can be used to create credentials and manage them.  This solution has integration with SES and Sendgrid/msg91 for sending emails and SMSs.

If one is not interested in UI layer of #MARK Studio, then they can use the [APIs available](https://api.demo.dhiway.com/docs/) too.

-------

We are also working on making our ledger and VC library work smoothly with Sunbird RC solutions, so one would get the same behaviour using Sunbird stack too. This is taking bit more time than anticipated.


CC: @smohan-dw @sankarshanmukhopadhyay

**Comment by suresh12** [comment]
Thanks everyone. We have reviewed Sunbird Admin Portal and Sunbird Issuance Portal(https://github.com/Sunbird-RC/sunbird-rc-issuance-ui) for ULP Implementation. We will extend those applications to to include additional functionalities needed for the above use cases.
